### ISSUE TITLE
Consider adding a "free rider" and "findpeers" options

### ISSUE THREAD

**Issue by Ayms** [mafintosh/torrent-stream#17]
Free rider: disable storing pieces (only mem[index] is used in memory) and disable sending 'have' to other peers, other peers can still see you of course but you minimize your participation to a torrent, the rational is for a project like Peersm where other peers are relaying data for you and then should not keep what they are relaying and neither advertise themselves. Small changes that I can provide if there is an interest.

Findpeer: there is a TODO comment in the code, I don't know exactly why the 10000 value was chosen but it causes in my case an average delay of 2mn before things start, maybe it could become an option (in the meantime routing tables are added???)

**Comment by dcerisano** [comment]
@Ayms  The bittorrent algorithm implements a Give-and-Take criterion (GT criterion) which is designed to discourage freeriders. Examples are "tit-for-tat" and the more modern "give-to-get" criteria.

Your mods just block uploading (sharing), which the GT criterion of the peers will detect. They will down-vote this peer and simply stop sharing with it.

Freeriding is Smith: every man for himself.
BitTorrent is Nash: all for one and one for all.

The two approaches are incompatible and the latter is known to lead to superior solutions.

**Comment by Ayms** [comment]
There is the theory and the practice, the bittorrent protocol does nothing to forbid freeriding, worse, torrent-live for example is not only not sharing but just does not respond to anything with the complete freerider option which is the default, why it is the default is explained in torrent-live repo or [torrent-live](http://www.torrent-live.org)

And this is working very well and why this is working is explained in the above links.

To understand maybe better the intent you can take a look at [Monitoring and blocking the bittorrent monitoring spies](https://mailman.stanford.edu/pipermail/liberationtech/2015-February/015102.html)

Complete freeriding is indeed a deviant behavior, but everybody knows that everybody is simply freeriding (ie not sharing after they get out of the swarm), and other deviant behaviors are vpn peers, or monitors, or popcorn stuff, etc

Now, should it be added to torrent-stream, I don't know, but I have made quite a lot of modifications since the initial post.

**Comment by dcerisano** [comment]
AFAIK, an upload rate of zero kills torrent downloads. Peers that allow freeriding constitute a separate swarm. Please understand the GT enabled peers will simply share among themselves exclusively.

**Comment by Ayms** [comment]
Yes, I did not say that freeriding was good (and morever total freeriding like torrent-live), it is explained in the repo why it is currently the default.

But I don't get very well what's your point regarding this issue, what do you propose here?

**Comment by gordonperlin** [comment]
Opting out the of GT criterion will cause the swarm to stop uploading. No give, no take.
Rate limiting requests from peers (throttling uploads)  value is all that is possible.
This satisfies the GT criterion, and prevents being isolated from the swarm.

**Comment by dcerisano** [comment]
The GT criterion is not satisfied if by a node that serves freeriders. Such a node is not a peer, and ironically is only detectable by attempting to freeride from it.

So effective peer detection (GT compliance) must:
- Detect freeriders
- Detect nodes that serve freeriders

**Comment by dcerisano** [comment]
![image](https://cloud.githubusercontent.com/assets/1641613/8391878/1343306e-1ca3-11e5-936e-bbcaef4b5790.png)

**Comment by Ayms** [comment]
What bittorrent clients use this from your standpoint (uTorrent?)

I have not studied this, but in practice totally freeriding (ie even not answering to anything) works very well, once torrent-live reaches the closest nodes, based on some criteria it chooses 20 peers in the swarm and download from them. Some peers apparently get pissed off that torrent-live's peer is not cooperating and disconnect, then torrent-live switches to another one, after some time torrent-live's swarm stabilizes and oscillates around 20 peers.

This works very well (even with small swarms), so if the most used bittorrent clients are implementing the anti freeriders features, it seems that this just does not work, again I did not check, but as far as I know popocorn time is freeriding too, I don't know to what extent.

**Comment by dcerisano** [comment]
Sounds like you have some data to post.
Would like to see a comparison of peer vs freerider performance with torrents of various swarm sizes (actual data).

Could you fill this in so your results can be verified?

    Size     Resolution    Seeders    Leechers     Peer (minutes)     Freerider (minutes)
    800 MB      720p       ~100        ~50      
    800 MB      720p       ~50         ~100      
    800 MB      720p       ~1000       ~500    
    800 MB      720p       ~500        ~1000     
    1.5 GB     1080p       ~50         ~100
    1.5 GB     1080p       ~100        ~50
    1.5 GB     1080p       ~500        ~1000
    1.5 GB     1080p       ~1000       ~500

**Comment by Ayms** [comment]
You did not answer the question (what bittorrent clients are supposed to implement the anti freeriders features?) and I don't really get your test case, you can try torrent-live if you want to get some statistics.

**Comment by dcerisano** [comment]
I did answer the question with a request for data to support your performance claim. 
It is not my claim to support.
Choking leeches is the basis of the peer-wire-protocol.

**Comment by Ayms** [comment]
Obviously you are just a complete idiot trolling this thread in order to try understanding what torrent-live does

**Comment by asapach** [comment]
Locking this issue to prevent further flame wars.
